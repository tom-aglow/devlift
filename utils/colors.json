{
  "white": "#FBFBFB",
  "lightGrey": "#E7E7E7",
  "mediumGrey": "#C2C2C2",
  "black": "#202020",
  "primary": "#12B4C2"
}
