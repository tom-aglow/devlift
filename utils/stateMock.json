{
  "todos": [
    { "id": 1, "text": "foo", "isCompleted": false, "listId": 0 },
    { "id": 2, "text": "bar", "isCompleted": true, "listId": 0 },
    { "id": 3, "text": "baz", "isCompleted": true, "listId": 1 }
  ],
  "lists": [
    { "id": 0, "title": "Personal" },
    { "id": 1, "title": "Movies to watch" }
  ],
  "inputValue": "",
  "editingId": 0,
  "openLists": [0]
}
