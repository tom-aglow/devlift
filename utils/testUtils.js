import React from "react";
import { mount, shallow } from "enzyme";
import { JSDOM } from "jsdom";

export const sel = id => `[data-test="${id}"]`;

export const setupGlobalObject = () => {
  const jsdom = new JSDOM("<!doctype html><html><body></body></html>");
  const { window } = jsdom;
  global.window = window;
  global.document = window.document;
};

export const TestHelper = class TestHelper {
  constructor(component, defaultProps) {
    this.component = component;
    this.defaultProps = defaultProps;
  }

  mountComponent(props = {}, isShallow = false) {
    const propsToUse = {
      ...this.defaultProps,
      ...props
    };

    if (isShallow) {
      return shallow(React.createElement(this.component, propsToUse));
    }

    return mount(React.createElement(this.component, propsToUse));
  }
};
