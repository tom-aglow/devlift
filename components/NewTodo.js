import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  StyleSheet,
  View,
  TextInput,
  Image,
  TouchableOpacity
} from 'react-native';
import RadioForm, {
  RadioButton,
  RadioButtonLabel
} from 'react-native-simple-radio-button';

import colors from '../utils/colors.json';

class NewTodo extends Component {
  state = {
    listId: 1,
    areChipsVisible: false
  };

  handleSubmit = () => {
    this.props.onAddItem({ listId: this.state.listId });
  };

  handleRadioButtonSelect = value => {
    const newState = Object.assign({}, this.state, { listId: value });
    this.setState(newState);
  };

  handleInputFocus = () => {
    const newState = Object.assign({}, this.state, { areChipsVisible: true });
    this.setState(newState);
  };

  handleInputBlur = () => {
    const newState = Object.assign({}, this.state, { areChipsVisible: false });
    this.setState(newState);
    this.props.onChange('');
  };

  renderInput() {
    const { inputValue, onChange } = this.props;

    return (
      <View style={styles.inputWrapper}>
        <TextInput
          value={inputValue}
          onChangeText={onChange}
          onSubmitEditing={this.handleSubmit}
          onFocus={this.handleInputFocus}
          onBlur={this.handleInputBlur}
          placeholder="I want to..."
          blurOnSubmit={false}
          returnKeyType="done"
          style={styles.input}
        />
        <TouchableOpacity onPress={this.handleSubmit}>
          <Image source={require('../img/button_add.png')} />
        </TouchableOpacity>
      </View>
    );
  }

  renderListChips() {
    const chips = this.props.lists.map(({ title, id }) => {
      const isActive = this.state.listId === id;

      return (
        <RadioButton labelHorizontal={true} key={id}>
          <RadioButtonLabel
            obj={{ label: title, value: id }}
            index={id}
            labelHorizontal={true}
            onPress={this.handleRadioButtonSelect}
            labelStyle={[
              styles.radioButton,
              isActive && styles.radioButtonActive
            ]}
            labelWrapStyle={[
              styles.radioButtonWrapper,
              isActive && styles.radioButtonWrapperActive
            ]}
          />
        </RadioButton>
      );
    });

    return (
      <View>
        <RadioForm formHorizontal={true}>{chips}</RadioForm>
      </View>
    );
  }

  render() {
    const { areChipsVisible } = this.state;

    return (
      <View style={styles.header}>
        {this.renderInput()}
        {areChipsVisible && this.renderListChips()}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  header: {
    paddingHorizontal: 10,
    flexDirection: 'column',
    backgroundColor: colors.white,
    paddingBottom: 5,
    paddingTop: 15,
    elevation: 4
  },
  inputWrapper: {
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    marginBottom: 15
  },
  input: {
    flex: 1,
    height: 30,
    fontFamily: 'OpenSans-Light',
    borderColor: colors.lightGrey,
    borderWidth: 1,
    borderTopRightRadius: 15,
    borderBottomRightRadius: 15,
    borderTopLeftRadius: 15,
    borderBottomLeftRadius: 15,
    paddingHorizontal: 15,
    marginRight: 20
  },
  radioButtonWrapper: {
    height: 30,
    marginLeft: 10,
    borderColor: colors.primary,
    borderWidth: 1,
    borderTopRightRadius: 15,
    borderBottomRightRadius: 15,
    borderTopLeftRadius: 15,
    borderBottomLeftRadius: 15,
    paddingHorizontal: 15,
    paddingVertical: 5
  },
  radioButtonWrapperActive: {
    backgroundColor: colors.primary
  },
  radioButton: {
    color: colors.primary,
    fontFamily: 'OpenSans-Light',
    paddingLeft: 0
  },
  radioButtonActive: {
    color: colors.white
  }
});

NewTodo.propTypes = {
  inputValue: PropTypes.string.isRequired,
  lists: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired
    })
  ),
  onChange: PropTypes.func.isRequired,
  onAddItem: PropTypes.func.isRequired
};

export default NewTodo;
