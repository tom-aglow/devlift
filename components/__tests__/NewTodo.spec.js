import NewTodo, { styles } from '../NewTodo';
import { TestHelper, sel, setupGlobalObject } from '../../utils/testUtils';
import stateMock from '../../utils/stateMock.json';

const defaultProps = {
  inputValue: stateMock.inputValue,
  lists: stateMock.lists,
  onChange() {},
  onAddItem() {}
};

const helper = new TestHelper(NewTodo, defaultProps);

beforeAll(() => {
  setupGlobalObject();
});

it('renders correctly', () => {
  const wrapper = helper.mountComponent({}, true);
  expect(wrapper).toMatchSnapshot();
});

it('shows and hide list chips when user focus or blur on the input', () => {
  const wrapper = helper.mountComponent({}, true);

  let chips = wrapper.find(sel('list-chip'));
  expect(chips).toHaveLength(0);

  const input = wrapper.find(sel('new-todo-input'));
  input.simulate('focus');
  chips = wrapper.update().find(sel('list-chip'));
  expect(chips).toHaveLength(stateMock.lists.length);

  input.simulate('blur');
  chips = wrapper.update().find(sel('list-chip'));
  expect(chips).toHaveLength(0);
});

it('selects correct list when user clicks on a chip', () => {
  const wrapper = helper.mountComponent({});
  const input = wrapper.find(sel('new-todo-input')).first();
  const chipIndex = 1;
  const listId = stateMock.lists[chipIndex].id;

  input.simulate('focus');

  let chipTwo = wrapper.find(sel('list-chip')).at(chipIndex);
  chipTwo.props().onPress(chipTwo.props().obj.value);

  chipTwo = wrapper
    .update()
    .find(sel('list-chip'))
    .at(chipIndex);

  expect(chipTwo.props().labelWrapStyle).toContain(
    styles.radioButtonWrapperActive
  );
  expect(wrapper.state('listId')).toBe(listId);

  const chipOne = wrapper.find(sel('list-chip')).at(0);
  expect(chipOne.props().labelWrapStyle).not.toContain(
    styles.radioButtonWrapperActive
  );
});
