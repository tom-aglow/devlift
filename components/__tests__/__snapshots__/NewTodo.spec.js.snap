// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FBFBFB",
      "elevation": 4,
      "flexDirection": "column",
      "paddingBottom": 5,
      "paddingHorizontal": 10,
      "paddingTop": 15,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginBottom": 15,
        "width": "100%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      blurOnSubmit={false}
      data-test="new-todo-input"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="I want to..."
      returnKeyType="done"
      style={
        Object {
          "borderBottomLeftRadius": 15,
          "borderBottomRightRadius": 15,
          "borderColor": "#E7E7E7",
          "borderTopLeftRadius": 15,
          "borderTopRightRadius": 15,
          "borderWidth": 1,
          "flex": 1,
          "fontFamily": "OpenSans-Light",
          "height": 30,
          "marginRight": 20,
          "paddingHorizontal": 15,
        }
      }
      value=""
    />
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
    >
      <Image
        source={
          Object {
            "testUri": "../../../img/button_add.png",
          }
        }
      />
    </TouchableOpacity>
  </View>
</View>
`;
